graceful reconfiguration v1 design doc (#26785)
MaterializeInc/materialize: doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

 - The action will be cancellable by canceling the query.

### V2 Scope
 - Provide a Hydration-based reconfiguration cleanup scheduling.
 - The DDL will be able to run in the background.
 - The action will be cancelable or able to be preemptively finalized.

## Out of Scope
 - Interactions with proposed schedules, ex: auto-scaling/auto-quiescing

## Solution Proposal

### Summary:
This feature will introduce new SQL in `ALTER CLUSTER` which will create new
replicas matching the alter statement; however, unlike existing alter mechanisms,
we will delay the cleanup of old replicas until the provided conditions are met.

The SQL will need to define a type of check to be performed, parameters for that
check, and whether or not to background the operation.

Suggested syntax:

```sql
ALTER CLUSTER c1 SET (SIZE 'small') WITH ( 
  WAIT = FOR <interval>,
  BACKGROUND = {true|false},
)
```

```sql
ALTER CLUSTER c1 SET (SIZE 'small') WITH ( 
  WAIT = UNTIL CAUGHT UP (
    TIMEOUT = <interval>,
    ON TIMEOUT = {CONTINUE | ABORT}
  )
  BACKGROUND = {true|false},
)
```

The two types of checks, in agreement with the V1 and V2 scopes respectively, are: 

>  - `FOR` which waits for a provided duration before failing and aborting.

Reviewer: > Wait why does it fail and abort? I thought `FOR` waited then returned success, regardless of the status or hydration state of the new replicas.

Agreed!

>  - `FOR` which waits for a provided duration before failing and aborting.

Author: Yeah... this is just wrong, I think conflated a timeout with for..

WAIT FOR will wait for a duration and proceed with finalization.

 - `UNTIL CAUGHT UP` which will wait for a "catch up" mechanism to return true
   or for a timeout to be met. It will roll back or forward depending on the value
   provided to `ON TIMEOUT`.

The SQL will kick off some action in `sequence_alter_cluster_managed_to_managed`
which will result in new reconfiguration replicas being created, new entries
placed `mz_internal.mz_pending_cluster_replicas` and a `ReconfigureCluster`
task being created which will ensure the reconfiguration finishes.

To keep track of ongoing reconfigurations we will introduce a new
`mz_internal.mz_pending_cluster_replicas` table which holds the state of

> reconfiguration replicas. This table will include all configuration parameters

Author: yeah, this has been removed, but should have been clarified.. It would only cover an internal and recoverable failure such as envd crash in the v2 scenario.

required to restart a reconfiguration task. It will be in charge of resuming or
cleaning up a reconfiguration.

All replicas will be billed during reconfiguration (except explicitly unbilled replicas).

### Definitions:
 - Active Replica: These replicas should be actively serving results and are
   named r1, r2, r3, etc. If a reconfiguration is ongoing, they will match the
   prior version of the configuration. If a reconfiguration is not ongoing they
   should be the only replicas.
 - Reconfiguration Replica: These replicas only exist during a reconfiguration,

>    their names will be suffixed with `reconfig-`. They will move to active

Reviewer: (See comment below about using a `-pending` suffix instead: https://github.com/MaterializeInc/materialize/pull/26785/files#r1606149023)

    clusters with reconfiguration replicas.
  - Must be run on managed clusters, and cannot be used to alter a cluster from
    managed to unmanaged.

#### Replica Naming
Active replicas will continue with the same naming scheme (r1, r2, ...)
Reconfiguration replicas will use the `reconfig` prefix; ex `reconfig-r3`,
`reconfig-r4`, which will be removed once they are moved to active.

#### Interactions with other schedules:

__V1__
We will choose to initially disallow any reconfiguration of a cluster
with a schedule.

__V2__
For refresh schedules, we will only allow graceful reconfiguration on clusters
whose refresh schedule is "ON", we would also need to prevent the clusters from being
turned off until the reconfiguration is finished.

__Future Schedules__
Some proposed schedules, auto-scaling, for instance, may want to use graceful
reconfiguration under the hood. These would need specific well-defined
interactions with user-invoked graceful reconfiguration.

#### Catalog Changes
We will need to create a new table `mz_internal.mz_pending_cluster_replicas`

| column     |  type |
| ---------- | ----- |
| replica_id | str   |

> | cluster_id | str   |

Reviewer: replica IDs are unique across clusters, so you only need the replica ID; better to leave the job of relating a replica ID to a cluster ID to the `mz_cluster_replicas` table

```suggestion
```

Author: This table has been removed. @maddyblue and I had a good conversation late last week and think we can get away without a tracking table, definitely for v1, and likely for v2.

| timeout    | int   |
| wait_type  | str   |
| background | bool  |
| start_time | int   |
| finished   | bool  |
| continue_on_timeout | bool |

> This table will be responsible for holding the state of the graceful replication task.

Reviewer: This is a very good point. Perhaps this would be much cleaner if we modeled specifically the task's state in a table called `mz_pending_cluster_alterations`? One row per outstanding alteration.

(Sorry, Justin — I feel like I suggested the one-row-per-replica idea on Slack.)

Each reconfiguration replica will have an entry in `mz_epnding_cluster_replicas`.

#### Reconfiguration Task

> The reconfiguration task will perform the following actions:

Reviewer 2: What is the task exactly? A tokio task? A coordinator message that gets sequenced and can possibly send other coord messages?

Author: Addressed in new update, and discussed in huddle, but I'll do a quick outline here.

Initially the thought was a tokio task, however, it seems like it'll be cleaner to do this through a cluster pending reconfiguration message handler that performs the check, finalizes the reconfiguration and returns to the user.

We'll have to adjust the sequencer to know it shouldn't retire the connection if it's a graceful reconfiguration.

1. Monitor the `wait` condition
The `wait` condition provided in the DDL will be polled by this task, once the
condition is met the replicas will be marked as finished and the task will move
on to finalization.

2. Finalize the reconfiguration
Finalizing the reconfiguration will occur once the `wait` condition has been
met. During this phase, the following steps will occur.
- reconfiguration replica promotion to active
- cluster configuration update to the new desired state

> - replicas with the previous configuration will be removed

Author: I'm just using catalog::Op for this, the drop op will be the first to run, this is pretty much how we're doing sequence_alter_cluster now, if this doesn't seem safe I'm happy to take a look at some additional error handling to ensure we've cleaned up the pending replicas?

https://github.com/MaterializeInc/materialize/pull/27904/files#diff-872e968ccc9dcf8e0e9e112062fa1918cfe5fab301cf701b9e5c91361fd62aabR307-R324

cluster config matches the original version.

#### Cluster Sequencer

The following roughly defines the new logic for the cluster sequencer when
performing an alter cluster on a managed cluster.

`mz_sql::plan::AlterClusterOptions` will have additional fields added to keep track of
the mechanism being used to perform the alter `AlterClusterMechanism::{Default|Duration|UntilCaughtUp}`.

If we see a non-default mechanism we'll ensure that the cluster is managed
and pass the mechanism to `sequence_alter_cluster_managed_to_managed`.
This function will need to be updated to either perform a
`sequence_alter_cluster_managed_to_managed_disruptive` or
`sequence_alter_cluster_managed_to_managed_graceful` operation.

The `disruptive` version will perform the current delete-before-create mechanism.
The `graceful` version will perform the following actions:
1. Validations
  1. validate that the cluster is not undergoing a reconfiguration
  2. validate the cluster is managed
  3. validate the cluster has no sources and now sinks
  4. the cluster cannot have a refresh schedule (v1)
2. Creating the reconfiguration replicas

> 3. Transactionally creating reconfiguration replicas and inserting the replicas into

Reviewer 2: "creating reconfiguration replicas" is both here and in the previous step.

   the `mz_pending_cluster_replicas` table.
4. Starting a blocking (or backgrounded) reconfiguration task 

> #### Recoviner from Environmentd Crash

Reviewer 2: ```suggestion
#### Recover from Environmentd Crash
```

Initially, we will recover from an environmentd crash by removing all

> reconfiguration replicas and clearing the `mz_pending_cluster_replicas` table,

Reviewer 2: I don't think the replicas need to be explicitly removed. On startup coord announces to compute the desired replica state, and compute complies including removing anything that exists but it was not told about (please verify with the compute team, but I'm like medium confident about this).

to ensure no active reconfigures are occurring. This is because
we will treat any disruption in connection as a failed DDL which will need to
be aborted.

> For V2, for each cluster, we will want to inspect whether an active

Reviewer 2: This is less a v2 thing and more a `BACKGROUND = true` thing yeah? Can it talk about background instead of v2 then? I kinda think since background isn't fully defined here, we should just omit this altogether maybe with a note somewhere saying the background or v2 stuff will be in a later design doc.

Author: I'm guessing the rollout of v1,v2 is actually

1 support for `WAIT FOR` foreground
2 `support for WAIT UNTIL` foreground
3 support for  background

For hydration-based scheduling the `until caught up` syntax will be used.

### Reconfiguration replica naming:
Is it still useful to have reconfiguration replica naming differences if we also have a table tracking those
replicas? It seems like we could get the list just as easily with a join and skip a step in the reconfiguration task.

### Should the reconfiguration task create the replicas
Currently, I have the reconfiguration task set to be in charge of the wait check and the finalization, but not
the reconfiguration replica creation nor the insertion into `mz_cluster_pending_replicas`. This seems like a reasonable
separation, as if we cannot do these two things we should fail and not background a task.

### Do we want a more generic jobs table?
The proposed mechanism is for a sort of one-off very specific job engine. Assuming we
want a more robust job engine in the future it may make sense to take baby steps towards that here.
For instance, we could introduce an `mz_job` table that stores all information about a ReconfiguratTask. This
could be used on environmentd restart to re-initialize the task or perform cleanup for the task if the envd restart

> would've caused the task to fail. I haven't thought this through much.

Reviewer: I'm extremely 👍🏽 on doing something one off for altering clusters (i.e., the design you've proposed), vs building a more generic jobs engine. CockroachDB has a generic jobs engine, and it's not terribly clean; different types of jobs have significantly different characteristics.

       - if new config != prior or current
         -  drop the reconf replicas, and create new reconf replicas.
            update the reconfiguration deadline as needed
     - if delay = false
       - follow the old logic of dropping and re-creating, drop all replicas,
         including reconfig replcias, reset the cluster schedule.
```
CONCERN:
Allowing delay=false and a timeout=0 provides two similar mechanisms of
using the old behavior. I think this is what we want as it ensures the old
replicas are destroyed before the new ones are created which has billining
implications. If we were to always rely on the reconfiguration_finalization
we may charge for both sets of replicas for some additional time if the
coordinator has a stall. If we're not concerned about this, always using
reconfiguration_finalization may be better as it reduces the code paths.
```

## Open Questions

### How do we handle multiple schedules?
If this feature is introducing multiple schedules we'll need some rules governing the interaction between schedules.
There's world where each schedule needs to consider it's interaction with all other schedules and that seems kind of

> scary.

Reviewer 3: I think there could be a division of schedules into 2 categories:
1. Schedules that turn clusters On/Off.
2. Schedules that affect the set of replicas of a cluster, _if the cluster is On_.

The interaction among all the schedules in the 1. category are already handled by the current code. (We wait for all schedules to have their say about a cluster before touching a cluster, and the cluster will be On if at least one schedule says that it should be On.)

The interaction between 1. and 2. would be:
- If 1. decided that the cluster is Off, then 2. has nothing to do, because there are no replicas.
- If 1. decided that the cluster is On, then 2. can decide on the exact set of replicas.

This leaves open the interactions among different schedules in category 2. What would be another example of a schedule in 2.? Maybe auto-scaling? I guess that would need to have some custom-defined interaction with graceful reconf timeout schedule.

### How does a multi-subscriber coord work with schedules?
We definitely don't want two schedule decisions to be acting at the same time. Currently this isn't

> an issue, but it seems a bit scary. Is this a concern?

Reviewer 3: I'd guess that stuff like altering clusters will always be a centralized thing, and only 1 instance of  `handle_scheduling_decisions` would be allowed to run at a time. (But I'm not an expert on PlatformV2 stuff.)

  - Unbilled replicas do not count towards reconfiguration and are not affected
    by reconfiguration.

#### Limitations:
  - Cannot be used on clusters with sources, sources should not be created on
    clusters with reconfigure replicas.

#### Replica Naming
Active replicas will continue with the same naming scheme (r1, r2, ...)
Reconfiguration replicas will use the `reconfig` prefix; ex `reconfig-r3`,
`reconfig-r4`, which will be removed once they are moved to active.

#### SQL
Introduce new SQL to define a delay duration for alter cluster.

`ALTER CLUSTER c1 SET (SIZE 'small') WITH ( CLEANUP TIMEOUT  5 minutes );`

This will alter the cluster and create new cluster replicas, but will not
remove existing replicas. Instead, it will set the schedule for the cluster
to `ClusterSchedule::Reconfigure` and set the deadline to `now + the provided
duration`. When CLEANUP TIMEOUT is not provided the cluster will behave as it
previously did, immediately tearing down and then creating the new replicas.

Initially, all reconfigurations will wait for the full timeout; however, when we
move to a hydration/parity cleanup detection mechanism, the reconfiguration may

> finish before the timeout. The same SQL can be used in both cases.

Author: I really like this. Initially I had a 
`WITH [<condition name>] CLEANUP TIMEOUT = <interval>` 
but  `UNTIL HYDRATED WAIT FOR 10 minutes` sounds way better.

@maddyblue had questioned whether we need to expose the `HYDRATION`/`CAUGHT UP` mechanism at all. I think this is a super valid point. 
- Would a user care about the mechanism being used to trigger cleanup? 
- Would a user ever want to wait for the full duration when we have a mechanism to trigger the cleanup early?

If we don't want to do provide this dials, we could consider something like
```
WITH ( CREATE BEFORE DELETE, MAX OVERLAP = 10m )
```
This is generated from the part of my brain that works in terraform/pulumi, but.. those are also tools that are likely to be directly interacting with creating/resizing clusters.

> finish before the timeout. The same SQL can be used in both cases.

Reviewer 4: Agree with Nikhil that `CLEANUP TIMEOUT` is too specific. Snowflake uses a syntax similar to his proposal for warehouse resizing: [`WAIT_FOR_COMPLETION`](https://docs.snowflake.com/en/sql-reference/sql/alter-warehouse#label-alter-warehouse-wait-for-completion).

> finish before the timeout. The same SQL can be used in both cases.

Author: > No, I don't think so—I think that's why you want FOR ... and UNTIL HYDRATED to be mutually exclusive options.

I'm not entirely convinced that we want `FOR <duration>` and `UNTIL CAUGHT UP` to be mutually exclusive. It's possible we do, but we would need to be pretty confident in our `UNTIL CAUGHT UP` mechanism if we go that route. My preference would be to provide the user the knob to set a hard limit on the overlap duration, potentially failing forward.

> Agree with Nikhil that CLEANUP TIMEOUT is too specific. Snowflake uses a syntax similar to his proposal for warehouse resizing: [WAIT_FOR_COMPLETION](https://docs.snowflake.com/en/sql-reference/sql/alter-warehouse#label-alter-warehouse-wait-for-completion).

I'm not sure [WAIT_FOR_COMPLETION](https://docs.snowflake.com/en/sql-reference/sql/alter-warehouse#label-alter-warehouse-wait-for-completion) is quite what we're doing.
The usage description I see is: `To block the immediate return of the ALTER WAREHOUSE command until the resize is complete, add the WAIT_FOR_COMPLETION parameter.`, which is really about changing the return strategy of the query not the underlying mechanism of the `alter`.  We'd discussed having a `BACKGROUND=true|false` option as part of this, and that is directly comparable to `WAIT_FOR_COMPLETION`.

2. the reconfigue config (what is in the catalog)

First scenario, the cluster being altered has no `Reconfigure` `cluster_schedule`:
If cleanup_timeout is None:
 - use the old mechanism ( drop and replace )
If the new alter statement provides a `cleanup_timeout`: 
 - check for sources/sinks,
 - update the catalog with the changes and add a `Reconfigure` cluster schedule
 - deploy new reconfigure replicas

Second scenario, the cluster being altered is undergoing a reconfigure:
If cleanup_timeout is None,
 - Use the old mechanism - Drop and replace including all reconfigure replicas and remove the schedule.
If the new alter statement provides a `cleanup_timeout`:
 - If the new config matches the current config
   - bump the timeout
 - If the new config doesn't match the current config
   - check for sources/sinks
   - update cluster catalog with new schedule and config
   - drop reconfigure replicas
   - launch new reconfigure replicas

Note:
An `ALTER CLUSTER` statement that is run twice (double shot), either needs to
push out the timeout of an ongoing reconfiguration or needs to entirely replace
the existing reconfigure schedule dropping and recreating any reconfigure

> replicas. I believe the former is the more useful behavior.

Reviewer: I'd like to propose an alternative v1: just block the `ALTER CLUSTER` command until the cutover is complete. I [mentioned this briefly on the issue](https://github.com/MaterializeInc/materialize/issues/20010#issuecomment-2058302174):

> Should the ALTER CLUSTER statement block until the reconfiguration is complete? This is the "correct" behavior from the PostgreSQL perspective. For example, CREATE INDEX and ALTER TABLE both block until the potentially very slow operation is complete. This has the appealing property of allowing us to report failures to complete the operation back to the user. It also seems a good bit simpler to build, because we don't need to build any job scheduling infrastructure.

I think it'd be muuuuch simpler to start like this, since you basically wouldn't need any durable state to record that the alteration was in progress. Here's how it'd work:

1. `ALTER CLUSTER` command is planned normally, resulting in a call to `sequence_alter_cluster` on the coordinator main loop.
2. If `ALTER CLUSTER` command involves a reconfiguration, add the new replicas with the `-pending` suffix (and `pending = true` in the `mz_cluster_replicas` table) in a catalog transaction. In the same catalog transaction, ensure there are no existing pending replicas—if there are, send back an error, as another `ALTER CLUSTER` operation on the same cluster is already in progress. (In the future, we might prefer to have subsequent `ALTER CLUSTER` operations just _block_ until the earlier operation completes, but that's more work to rig up.)
3. Off of the coordinator main loop, wait out the `WAIT` policy.
4. Once the `WAIT` policy completes, issue another catalog transaction (on the coordinator main thread) that drops the old replicas as removes the `pending` bit from the new replicas.
5. Report success back to the user.

If the client goes away while the `ALTER` is running, we can have the coordinator abort the operation: just drop all the outstanding pending replicas. 

I don't think you'd need to interact with the cluster scheduler at all, other than to make sure that the schedule policy wasn't `ON REFRESH`. The interaction between `SCHEDULE = ON REFRESH` seems real tricky, and since `REFRESH EVERY` views are generally for cold/warm data rather than hot data, the lack of graceful cluster reconfiguration seems quite tolerable.

There's only one subtle bit: what if `envd` crashes in the middle of handling an `ALTER CLUSTER` command? E.g., say the `ALTER CLUSTER` is being executed during the maintenance window. I think the answer is pretty clean: just have `envd` drop all pending replicas when it boots. We _know_ that if we have pending replicas, the old `envd` 1) was in the middle of processing an `ALTER CLUSTER` command and 2) returned a "connection closed" error or somesuch that did not indicate success, and so we can simply drop the pending replicas and pretend the alteration never happend.

And, if blocking `ALTER CLUSTER` statements _does_ turn out to be a pain point, we can roll out support for `ALTER CLUSTER ... (BACKGROUND = TRUE)` in some future release of Materialize.

Author: I have some concerns about implementing non-blocking as follow up. As someone who regularly deals with cloud infra I find it much harder to work with things that must be handled over a single connection. Networks are flaky, laptops go to sleep, and connection timeouts threaten longer lived calls. 

This might work fine over PGWIRE where it's not abnormal to have long running queries (10s of minutes), but HTTP calls via the console may not be as resilient here. The connections themselves may be fine if they don't hit timeouts, but may find ourselves fighting with user interactions with long lived connections in the browser.

it never succeeded with the new size.

Benefits
  - Should be relatively easy to write this, as the expectation for any schedule
    adjustment should be idempotent, ensure the cluster/replicas are in the
    correct state after the decision action has finished.
Downsides
  - The interaction with `REFRESH EVERY` here is still weird. One might downsize
    during the active period of `REFRESH EVERY`, but refresh finishes before the
    reconfiguration finishing and the user will not know that the smaller size
    will OOM or OOD(out of disk) on the next run because the reconfigure size
    never had to finish.

### Allow cancelation?
We could additionally add a prior config field in `ClusterSchedule::Reconfigure`
that stores the `prior_config`. This would be the pre-reconfigure configuration
and could be used to cancel an upgrade. This would be done by running an
`ALTER CLUSTER` statement that does not provide a `CLEANUP TIMEOUT`, and whose
configuration matches the `prior_config`. This seems somewhat convoluted and
I don't know that allowing for cancellation is necessary as long as we allow for
adjustment of the current reconfigure timeout, and we allow for overwriting the

> current reconfiguration entirely.

Author: I still need to update the doc based on the discussion earlier today. Here's where I think we're leaning.. If the user believes the new smaller size will not work they would cancel the existing query being run assuming they are running a blocking query. If they are running the query in the background they would issue a cancel command like `cancel alter cluster <x>`.  In either case the existing reconfiguration replicas disappear and the actives state active.

> current reconfiguration entirely.

Reviewer: That's my understanding too!

`sequence_alter_cluster_managed_to_managed_disruptive` or
`sequence_alter_cluster_managed_to_managed_graceful` function.

The following are validations to run if the `AlterClusterPlan` contains
an alter_mechanism.
1. validate the cluster is managed
2. validate that the cluster is not undergoing a reconfiguration
3. validate the cluster has no sources and now sinks
4. the cluster cannot have a refresh schedule (v1)

`sequence_alter_cluster_managed_to_managed_disruptive` will  perform the current
delete-before-create mechanism.

`sequence_alter_cluster_managed_to_managed_graceful`  will still apply
all limits and validity checks to the `new configuration`, but will not
perform deletes of the active replicas, nor commit the new config to the
catalog. Instead, it will ask the compute controller to create new replicas,
add a new entry in  `pending_cluster_reconfigurations`,  then kick off a
`ClusterReconfigureCheckReady` message.

The state for the reconfiguration will largely be stored in a `PendingClusterReconfiguration`
struct.

```rust
struct PendingClusterReconfiguration {
 ctx: ExecuteContext,
 // used to determine if the message_handler
 // should wait, or finalize

>  is_ready: bool

Reviewer 2: Possibly this is not needed. It might be able to be an impl method on the struct.

Author: What do you think about having the `handle_cluster_reconfiguration` message handler  always validate the check and re-sen the message if it's false. We wouldn't need to keep track of is ready at all.
Ex:
```
if the start_time.elapsed > wait_time {
   //finalize
 } else {
         mz_ore::task::spawn(|| "", async move {
            // some wait condition
            tokio::time::sleep(wait_time - start_time.elapsed).wait;
            let _ = internal_cmd_tx.send(Message::ClusterReconfigurationReady(connid));
        });
 } 
 ```
 
 This makes more sense in the `WAIT UNTIL CAUGHT UP` case, where, if not caught up, it would just send the message in 3 seconds.

Reviewer 2: Yes, this is good.

Author: coool updated!

 // needed to determine timeouts
 start_time: Instant,
 // the plan will hold the cluster_id, config, mechanism, etc...
 plan: AlterClusterPlan,
 // idempotency key will need to be part of the  message and will need to be
 // validated to ensure we're looking at the correct plan, in case we
 // miss a cancel and a new entry was put in pending_cluster_reconfiguration.

>  idempotency_key: i64,

Reviewer 2: maybe needs to be a uuid

}

```

In order to cancel we'll keep a map, `pending_cluster_reconfigurations`,  of
session `connectionIds` to `PendingClusterReconfigurations`.

*Currently due to limitations in sources and sinks, we will need to maintain
the disruptive alter mechanism. Once multiple replicas are allowed for sources
and sinks we can remove the disruptive mechanism and default the wait to `WAIT=For 0 seconds`.

#### Handling PendingClusterReconfigure Messages

The responsibilities of this operation will be:
1. checking for cancellations
2. checking for the `WAIT` condition
3. finalizaing the reocnfigure
4. returning to the user

__Finalize the Reconfiguration__
Finalizing the reconfiguration will occur once the `wait` condition has been
met. During this phase, the following steps will occur.
- reconfiguration replica promotion to active
- cluster configuration update to the new desired state
- replicas with the previous configuration will be removed
- replicas for the cluster will be removed from the `mz_pending_cluster_replicas` table

3. Reverting the reconfiguration

> On failure or timeout of the reconfiguration, this task will revert to the state

Reviewer 2: Reverting due to timeout is only a v2 feature right? Timeout in v1 should not revert, but proceed.

prior to the `ALTER CLUSTER` DDL. It will remove reconfigure replicas, and ensure the
cluster config matches the original version.

__Handling Cancelation__
When a cancel event is read we'll check to see if the active connection has a value

> in a `pending_cluster_reconfiguration` BTreeMap, and if so remove it.

Reviewer 2: ```suggestion
in a `pending_cluster_reconfiguration` BTreeMap, and if so remove it and clean up leftover state.
```

Worth saying that we need to clean up some other stuff too.

   prior version of the configuration. If a reconfiguration is not ongoing they
   should be the only replicas.
 - Reconfiguration Replica: These replicas only exist during a reconfiguration,
   their names will be suffixed with `-pending`. They will move to active
   once the reconfiguration is finalized. These will be stored in the
   `mz_pending_cluster_replicas` table.
 - Reconfiguring Cluster: any cluster with one or more replicas in the
   `mz_pending_cluster_replicas` table
 - Replica Promotion, during finalization of an alter cluster, pending replicas will be "promoted".
   This entails changing the pending value of the replica to false, and removing the '-pending'
   suffix.

### Guard Rails
- Only one reconfiguration will occur at a time for a given cluster. Alter statements
  run against a reconfiguring cluster will fail.
- We must protect against creating a source/sink on a reconfiguring cluster and
  reconfiguring a cluster with sources/sinks. These clusters will need to use
  the non-scheduled mechanism (delete-before-create). This is due to a limitation
  where sources and sinks must only run on a single replica.

> - We may want to consider limiting timeout duration.

Reviewer 2: Solidy this wording: remove this line or come up with a specific proposal.

- If we are not backgrounding the reconfiguration the expectation will be that
  if the sql query does not successfully return, due to a crash or connection drop, the alter
  will not finalize and all reconfiguration replicas will cleaned up.

### Limitations:
  - Cannot be used on clusters with sources or sinks, sources should not be created on
    clusters with reconfiguration replicas.
  - Must be run on managed clusters, and cannot be used to alter a cluster from
    managed to unmanaged.

### Visibility
Reconfiguration replicas should be identifiable for purposes of reporting and

> visibility in the UI. We may be able to do this by creating a suffix `-pending`

Reviewer 2: Solidify the wording here to say what will happen instead of what may happen.

for reconfiguration replicas, alternatively, we could set a bool status on the
replica resource in the catalog.

### Replica Naming
Active replicas will continue with the same naming scheme (r1, r2, ...)
Reconfiguration replicas will use the `-pending` suffix; ex `r1-pending`,
`r2-pending`, which will be removed once they are moved to active.

### Implementation Details
We want to accomplish the following
 - interpret SQL to identify when to use graceful vs disruptive alters
 - selectively choose which mechanism is used to alter the cluster
 - provide a mechanism to cancel the operation

#### Catalog Changes

> We should add a `pending: bool` field to the `ClusterReplica` memory and durable

Reviewer 2: ```suggestion
Add a `pending: bool` field to the `ClusterReplica` memory and durable
```

catalog structs. This will be set to true for reconfiguration replicas and will

> help either recover or cleanup on environmentd crashes. We should also propagate
> this to `mz_cluster_replicas`.

Reviewer 2: ```suggestion
help either recover or cleanup on environmentd crashes. Propagate
this to `mz_cluster_replicas`.
```

#### Process
Below broken this process broken down into four components
1. Interperet the query to determine if we're doing graceful reconfiguration
2. run `sequence_alter_cluster_managed_to_managed_graceful`
3. wait and finalize
4. return to the user

*In v2, steps 3 and 4 may be swapped.

__Interpreting the Query__
This will use standard parsing/planning, the only notable
callout here is that we'll want `mz_sql::plan::AlterClusterOptions`
to contain an `alter_mechanaism` field which holds an optional

> `AlterClusterMechanism` which will look something like:

Reviewer 2: ```suggestion
`AlterClusterMechanism`:
```

```rust
enum AlterClusterMechanism {
  Duration{
    timeout: Duration
    ...
  },
  UntilCaughtUp{
    timeout: Duration,
    continue_on_timeout: bool
    ...
  },
}
```

__Cluster Sequencer__

> The following roughly outlines the new logic for the cluster sequencer when

Reviewer 2: ```suggestion
The following outlines the new logic for the cluster sequencer when
```

the current validations, used in both `sequence_alter_cluster` and
`sequence_alter_cluster_managed_to_managed`, but will kick off a new
`StageResult::Handle` when an `alter_mechanism` is provided. The task in this
handle will perform waits/validations and pass off the finalization to
a `Finalize` stage.

Additionally, we'll need to perform some extra validations to ensure the
the `alter_mechanism` and `AlterClusterPlan` are compatible. These are:
1. validate the cluster is managed
2. validate that the cluster is not undergoing a reconfiguration
3. validate the cluster has no sources and now sinks
4. the cluster cannot have a refresh schedule (v1)

The `sequence_alter_cluster_managed_to_managed` method will need
to be updated to apply the correct name/pending status for new replicas
and only remove existing replicas when no alter_mechanism is provided.

For V2:
We will need to implement a message strategy that kicks the work off
off of the `sequence_staged` flow. In this scenario, we will still create new
replicas in the foreground, but the wait/checks will be performed in a separate
message handle by emitting a `ClusterReconfigureCheckReady` message.

> The state for the reconfiguration will largely be stored in a `PendingClusterReconfiguration`

Reviewer 2: "largely" implies that there's other state. Where is it? Can you remove the word largely because it's actually all here?

struct.

```rust
struct PendingClusterReconfiguration {
 ctx: ExecuteContext,
 // needed to determine timeouts
 start_time: Instant,
 // the plan will hold the cluster_id, config, mechanism, etc...
 plan: AlterClusterPlan,
 // idempotency key will need to be part of the  message and will need to be
 // validated to ensure we're looking at the correct plan, in case we
 // miss a cancel and a new entry was put in pending_cluster_reconfiguration.
 idempotency_key: Uuid,
}
```

> In order to cancel we'll keep a map, `pending_cluster_reconfigurations`,  of

Reviewer 2: This is no longer true due to sequence_staged. This will be a new field in ConnMeta similar to its drop_sinks.

session `ClusterIds` to `PendingClusterReconfigurations`.

*Currently due to limitations in sources and sinks, we will need to maintain
the disruptive alter mechanism. Once multiple replicas are allowed for sources
and sinks we can remove the disruptive mechanism and default the wait to `WAIT=For 0 seconds`.

#### Handling PendingClusterReconfigure Messages

The responsibilities of this operation will be:
1. checking for cancellations
2. checking for the `WAIT` condition
3. finalizaing the reconfiguration
4. returning to the user

__Checking the `WAIT` Condition__
The message handler will perform all checks on every run. If the wait condition
has not yet been met, we will reschedule the message to be resent. For each
reschedule we'll spawn a task that waits some duration and sends a
`ClusterReconfigureCheckReady` message with the connection_id.
For `WAIT FOR` we can wait for the exact duration that would cause the next check to succeed.

> For `WAIT UNTIL CAUGHT UP` we can just wait for three seconds.

Reviewer 2: Add an open question about if there's a better way to do this, for example starting a subscribe and waiting off-thread for that to reach some condition.

__Finalize the Reconfiguration__
Finalizing the reconfiguration will occur once the `wait` condition has been
met. During this phase, the following steps will occur.
- Drop any non pending replicas that do not have a billed_as value.
- Finalize the pending replicas:
  - Rename to remove the `-pending` suffix.
  - Alter config to set `pending` to true.
- Update the cluster catalog to the new desired config.

> __Handling Cancelation__

Reviewer 2: This is no longer true.

When a cancel event is read we'll check to see if the active connection has a value
in a `pending_cluster_reconfiguration` BTreeMap, and if so remove it and clean up leftover state.

__v2 coniderations__
Reverting the reconfiguration
On failure or timeout of the reconfiguration, this task will revert to the state
prior to the `ALTER CLUSTER` DDL. It will remove reconfigure replicas, and ensure the
cluster config matches the original version.

> #### Recover from Environmentd Crash
> Initially, for v1, we will not recover. We will have to remove all `pending`
> replicas.

Reviewer 2: ```suggestion
#### Recover after environmentd restart
For v1, environmentd restarts will remove `pending` replicas from the Catalog during Catalog initialization.
This will cause the compute controller to remove any pending physical replicas during its initialization (this is already an existing feature).
```

For v2, we will need to store the AlterClusterPlan durably for all backgrounded reconfigurations.

> We will then need the coordinator to attempt to continue an ongoing recongiuration on restart.

Reviewer 2: There's no need to continue anything here! On restart the catalog will remove the non-pending replicas, promote the pending to non-pending, then proceed with compute controller initialization like normal which will cause the new desired state to take affect.

Author: Is that the behavior we want? I imagine if environmentd were to get restarted leading a small amount of downtime we wouldn't want to remove all non-pending fully hydrated replicas and promote replicas that have an unknown hydration state? I would expect the behavior we want to strive for here is to continue with the alter and only perform the finalization once the alter strategy condition is met.

Reviewer 2: My understanding is that on envd restart we always get a whole new set of replicas. The new envd does not hook in to the in-progress hydrating replicas, so they give us no benefit. Recommend verifying this behavior with the cluster team.

Reviewer 3: If envd is restarted but some clusterds are not restarted, we aim to not disturb these clusterds, and reuse them as much as possible. That is, the coordinator [tries](https://github.com/MaterializeInc/materialize/blob/9a1b5a8d24b6da9237fbf546f2f93628ed9d9185/src/adapter/src/coord.rs#L2602) to bootstrap timestamps in a way that existing dataflows on the running clusterds can be reused, and the clusterds [try](https://github.com/MaterializeInc/materialize/blob/5a35bda0ece9b45ea230933b21255e780fbfd2b1/src/compute/src/server.rs#L548) to recognize when compute commands that they receive from a new controller reflect already running things. See also `platform/reconciliation.md`.

Author: If the v2 mechanism is not based on rehydration but a parity of the frontiers between old and new replicas then we should be able to handle both the case that existing flows are kept alive and the case where full rehydration is required. In the former we'd wait for new replicas to catch up ,and would only finalize the alter once that had occurred. In the latter, since both sets of replicas had their dataflows restarted, we'd expect to have parity immediately.

# Graceful Reconfiguration of Managed Clusters

Associated issues/prs:
[#20010](https://github.com/MaterializeInc/materialize/issues/20010)
[#26401](https://github.com/MaterializeInc/materialize/pull/26401)

## The Problem
Reconfiguring a manage dcluster via `alter cluster...` leads to downtime. This
is due to our cluster sequencer first removing the old replicas then adding new
replicas that match the new configuration. The duration of downtime can be seen
as extending through the full period it takes to rehydrate the new replicas.

## Success Criteria
Provide a mechanism which allows user to delay deletion of old replicas for
some user defined duraiton while the new replicas spin up. This mechanism should
allow for a follow on feature, where teardown is delayed until the newly created
replicas have finished hydrating, or reach some parity with the replicas they
are replacing.

## V1 Scope

>  - Timeout based reconfiguration cleanup scheduling

Author: I guess I see the timeout as bounds you're setting on the cost of zero downtime during reconfiguration and hydration-detection as our way of helping reduce those costs.

>  - Timeout based reconfiguration cleanup scheduling

Reviewer: > Is this useful for users? It will still lead to unpredictable downtime. (Perhaps some of this has been discussed in places I'm not aware of but:) I'm interested in not building the timeout feature and only building the hydration-detection feature.

The `WAIT = FOR <interval>` is definitely not the ideal user experience, but I think we can ship it much faster than we can ship `WAIT = UNTIL CAUGHT UP`. No one has yet proposed a clear implementation for computing "is this replica caught up?", and I'd hate to block this work 

Maybe the right goal is to aim for getting `WAIT = FOR <interval>` merged and in "private preview", with the expectation that it's unlikely that folks will want to use it. But we can at least check our assumption there! And while we're getting feedback on `WAIT = FOR <interval>` we can try to shake out the question of computing "is this replica caught up?", and it should drop into place pretty cleanly with the foundation we lay with `WAIT = FOR< interval>`.

## V2 Scope
 - Hydration based reconfiguration cleanup scheduling

# Out of scope
 - Issues arising from multi-subscriber message handling.

## Solution Proposal

#### Summary:
This feature will introduce new SQL in `ALTER CLUSTER`
which will delay the cleanup of replicas until a specified timeout, or
until a hydration check is triggered. It will be possible to preempt the cleanup
by respecifying the statement with a TIMEOUT of 0.
Ex:
```sql

> ALTER CLUSTER c1 SET (SIZE 'small') WITH ( CLEANUP [AUTO DETECT] TIMEOUT 5 minutes );

Author: I would expect that there's a difference between these two statements
`ALTER CLUSTER c1 SET (SIZE 'small') ;`
`ALTER CLUSTER c1 SET (SIZE 'small')  WITH ( CLEANUP TIMEOUT 0 );`

The former should use the old mechanism of drop/create.
The latter should create a reconfigure that fires a decision the the next cycle of handle_cluster_scheduling. 

Reasonsing:
If you have an ongoing reconfiguration but you want to preemptively end it, setting the timeout to 0 could provide the mechanism if we're able to detect that there aren't differences and modify a reconfigure schedule.

Keeping the former syntax and functionality alive seems useful for clusters with sources/sinks as well as very cost sensitive protection against something like coordinator stalls or envd crash, or even a poorly timed reconfigure which occurs during a maintenance window. I could be convinced that this isn't a real concern and we should always do a scheduled cleanup except in the sink/source case which the cluster sequencer could figure out internally.

```

> Here, `AUTO DETECT` indicates we would automatically detect via some parity check that the

Author: yup, I like it.

new replicas are as hydrated as the prior replicas. For now a timeout must be specified.

The new syntax and mechanisms will be built to work with auto cleanup mechanisms
built around cluster hydration but the syntax for that and the mechanism are not
in scope for this document.

All replicas will be billed during reconfiguraiton (except explicitly unbilled replicas).

#### Definitions:
  - Active Replica: These replicas should be actively serving results and are
  named r1,r2,r3,etc... if a reconfiguration is ongoing they will match the
  prior version of the configuration. If a reconfiguration is not ongoing they
  should be the only replicas.
  - Reconfiguration Replica: These replicas only exist during a
  reconfiguration , their names will be suffixed with `reconfig-`. They will
  move to active once the reconfiguration is finalized.

#### Guard Rails

> - Only one reconfigure can occur at a time for a given cluster. If a cluster

Author: This is exactly what I was thinking. There's a small bit on the cluster sequencing portion on this flow, but I'll make it clear throughout that this will be an ovewrite of the existing reconfigure.

  undergoing reconfiguration is modified we should handle that in a way that
  supplants the ongoing reconfiguratin with a new reconfiguration, or cancels
  reconfiguration all together.

> - Protect against creating a source on a reconfiguring cluster, and

Author: Currently we deny adding replicas or increasing the replication factor if the cluster has a source. I believe we just don't support logical replication of sources.

  reconfiguring a cluster with a source.. these clusters will need to use the
  non schedule mechanism (delete-before-create).

````
Possible guard rail
- Limiting the timeout Duration. In theory one could set an arbitrarily long
  duration, it maybe be reasonable to introduce some maximum to the timeout
  duration.
````

#### Visisibilty
Reconfiguration replicas should be identifiable for purposes of reporting and visibility
in the UI. We may be able to do this by create a prefix "reconfig-" for
reconfiguration replicas, alternatively we could set a bool status on the
replica resource in the catalog.

### Details

#### Invariants:

>   - Only one reconfiguration will be allowed at a time for a cluster.

Author: yeah, I'm not really sure what the implications are of switching the cluster_schedule from a single value to a vec. 
Having two refresh schedules doesn't make much sense nor does having two reconfigure schedules. Perhaps it's more clear to say that the list of schedules must be unique with respect to schedule type and for reconfigure this will be resolved by overwriting the existing reconfigure.

replicas matching the alter statement; however, unlike existing alter mechanisms,
it will delay the cleanup of old replicas until the provided `WAIT` condition is met.

The SQL will need to define a type of check to be performed, parameters for that
check, and whether or not to background the operation.

Suggested syntax:

```sql
ALTER CLUSTER c1 SET (SIZE 'small') WITH (
  WAIT = <condition>,
  BACKGROUND = {true|false},
```

The two types of checks, in agreement with the V1 and V2 scopes respectively, are:
 - `FOR`, which waits for a provided duration before proceeding with finalization.
    ```sql
    ALTER CLUSTER c1 SET (SIZE 'small') WITH (
      WAIT = FOR <interval>,
      BACKGROUND = {true|false},
    )
    ```
 - `UNTIL CAUGHT UP`, which will wait for a "catch up" mechanism to return true
   or for a timeout to be met. It will roll back or forward depending on the value
   provided to `ON TIMEOUT`.
    ```sql
    ALTER CLUSTER c1 SET (SIZE 'small') WITH (
      WAIT = UNTIL CAUGHT UP (
        TIMEOUT = <interval>,
        ON TIMEOUT = {CONTINUE | ABORT}
      )
      BACKGROUND = {true|false},
    )

>     ```

Reviewer: I really like how this shook out!

All replicas will be billed during reconfiguration (except explicitly unbilled replicas).

__Cancelation__
For v1, all `ALTER CLUSTER` statements will block, and graceful reconfiguration
will only be cancelable by canceling the query from the connection that issued it.

For v2, we will introduce backgrounded reconfiguration which will need some
new semantics in order to cancel. Something like the following should suffice.
```sql

> CANEL ALTER CLUSTER <cluster>

Reviewer: Alternative strawman syntax for discussion: `ALTER CLUSTER ... CANCEL`.

of the active, we send a claim that the check condition has been met and proceed with
finalization of the reconfiguration.

__Potential challenges with using hydration status__:
It is difficult to know when "hydration" has finished because:
  - New data coming in may change the timestamp and lead to hydration status inaccuracies
  - New data-flows can be created ad-hock which will need to be "hydrated"
  - hydration_status errors related to Environmentd restart?
    (https://materializeinc.slack.com/archives/CTESPM7FU/p1712227236046369?thread_ts=1712224682.263469&cid=CTESPM7FU)
Some of these can be resolved by only looking at data flows that were installed
when the replica was created. It seems like comparing active and reconfiguration
replicas will likely work better than attempting to gauge when hydration has
finished for a given replica. It also may allow us to shorten the timeout if all
replicas for a cluster undergoing reconfiguration are restarted.

For hydration-based scheduling the `until caught up` syntax will be used.

### Reconfiguration replica naming:
Is it still useful to have reconfiguration replica naming differences if we also have a table tracking those

> replicas? It seems like we could get the list just as easily with a join and skip a step in the reconfiguration task.

Reviewer: How else will the replicas coexist? (Replicas need unique names.)

  - Rename to remove the `-pending` suffix.
  - Alter config to set `pending` to true.
- Update the cluster catalog to the new desired config.

__Handling Dropped Connections__
Dropped connections will be handled by storing a `clean_cluster_alters: BTreeSet<GlobalId>` in the `ConnMeta`.
We will use this along with a `retire_cluster_alter` function which can be called during connection termination.
`retire_compute_sync_for_conn` provides an example of this.

__Handling Cancelation__
The mechanism used for handling dropped connections can be used for cancelation.
`cancel_compute_sync_for_conn` provides an example of this.

__v2 coniderations__
Reverting the reconfiguration
On failure or timeout of the reconfiguration, this task will revert to the state
prior to the `ALTER CLUSTER` DDL. It will remove reconfigure replicas, and ensure the
cluster config matches the original version.

#### Recover after Environmentd Restart
For v1, environmentd restarts will remove `pending` replicas from the Catalog
during Catalog initialization. This will cause the compute controller to remove
any pending physical replicas during its initialization (this is already an

> existing feature).k

Reviewer 2: "k"
